infobind/cache.go

```diff
@@ -0,0 +1,215 @@
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/sqlexec"
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultBindCacheSize = 5
+
+// bindData store the basic bind info and bindSql astNode.
+type bindData struct {
+	bindRecord
+	ast ast.StmtNode
+}
+
+// bindCache hold a bindDataMap, key:origin sql hash value: bindData slice.
+type bindCache struct {
+	Cache map[string][]*bindData
+}
+
+// Handle hold a atomic bindCache.
+type Handle struct {
+	bind atomic.Value
+}
+
+// HandleUpdater use to update the bindCache.
+type HandleUpdater struct {
+	parser         *parser.Parser
+	lastUpdateTime types.Time
+	ctx            sessionctx.Context
+	globalHandle   *Handle
+}
+
+type bindRecord struct {
+	OriginalSQL string
+	BindSQL     string
+	Db          string
+	Status      int64
+	CreateTime  types.Time
+	UpdateTime  types.Time
+	Charset     string
+	Collation   string
+}
+
+// NewHandleUpdater create a new HandleUpdater.
+func NewHandleUpdater(handle *Handle, parser *parser.Parser, ctx sessionctx.Context) *HandleUpdater {
+	return &HandleUpdater{
+		globalHandle: handle,
+		parser:       parser,
+		ctx:          ctx,
+	}
+}
+
+// NewHandle create a Handle with a bindCache.
+func NewHandle() *Handle {
+	handle := &Handle{}
+	return handle
+}
+
+// Get get bindCache from a Handle.
+func (h *Handle) Get() *bindCache {
+	bc := h.bind.Load()
+
+	if bc != nil {
+		return bc.(*bindCache)
+	}
+
+	return &bindCache{
+		Cache: make(map[string][]*bindData, defaultBindCacheSize),
+	}
+}
+
+// LoadDiff use to load new bind info to bindCache bc.
+func (h *HandleUpdater) loadDiff(sql string, bc *bindCache) error {
+	recordSets, err := h.ctx.(sqlexec.SQLExecutor).Execute(context.Background(), sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rs := recordSets[0]
+	defer terror.Call(rs.Close)
+
+	chkBatch := rs.NewRecordBatch()
+	for {
+		err = rs.Next(context.TODO(), chkBatch)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if chkBatch.NumRows() == 0 {
+			return nil
+		}
+		it := chunk.NewIterator4Chunk(chkBatch.Chunk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record := decodeBindTableRow(row)
+			err = bc.appendNode(h.ctx, record, h.parser)
+			if err != nil {
+				continue
+			}
+
+			if record.UpdateTime.Compare(h.lastUpdateTime) == 1 {
+				h.lastUpdateTime = record.UpdateTime
+			}
+		}
+	}
+}
+
+// Update update the HandleUpdater's bindCache if tidb first startup,the fullLoad is true,otherwise fullLoad is false.
+func (h *HandleUpdater) Update(fullLoad bool) error {
+	var (
+		err error
+		sql string
+	)
+	bc := h.globalHandle.Get()
+
+	newBc := &bindCache{
+		Cache: make(map[string][]*bindData, len(bc.Cache)),
+	}
+
+	for hash, bindDataArr := range bc.Cache {
+		newBc.Cache[hash] = append(newBc.Cache[hash], bindDataArr...)
+	}
+
+	if fullLoad {
+		sql = fmt.Sprintf("select original_sql, bind_sql, default_db, status, create_time, update_time, charset, collation from mysql.bind_info")
```

Comment:
When do we write the bindCache in to tikv?

Reply:
When we add a new bind or drop a bind, we will write the bind info to tikv

Reply:
Is this an unfinished work and another PR will fix this?

Reply:
Yes！

Reply:
Yes, add bind info will be writen in next pr